CycloneDDS events executor support

mauropasse opened the pull request:
CycloneDDS with Events executor now works as FastRTPS but there's still work to do. 
But I rather do the rest of the work after this PR is merged, since it'd be useful to ask for guidance in the real irobot PR.

Subscriptions, clients, services and guard condition works fine, but: 
Guards conditions to be properly implemented I think requires a little extra work but better to consult it first.
Events, only entity that I don't think they work, also need some reworking and consultation.

Commit:
Events executor subscriptions support

Commit:
Events executor guard conditions support

Commit:
Add clients&services support

Commit:
Support events (still work to do)

Commit:
Ctest fixes

irobot-ros commented on rmw_cyclonedds_cpp/src/rmw_node.cpp:
```diff
@@ -473,16 +533,41 @@ extern "C" rmw_ret_t rmw_service_set_listener_callback(
-    "rmw_service_set_listener_callback: not supported (yet)");
-  return RMW_RET_UNSUPPORTED;
+  auto srv = static_cast<CddsService *>(rmw_service->data);
+  auto entity_to_listen = srv->service.sub->enth;
```
`auto` should be used when the dataype can be understood from the context, which is not the case here

mauropasse replied:
done in 010fb92

irobot-ros commented on rmw_cyclonedds_cpp/src/rmw_node.cpp:
```diff
@@ -449,22 +467,64 @@ extern "C" rmw_ret_t rmw_set_log_severity(rmw_log_severity_t severity)
+  user_callback_data_t * data = &(sub->user_callback_data);
+
+  if (user_data && subscription_handle && callback) {
+    data->callback = callback;
```
we should always update `sub->user_callback_data` regardless of whether they are valid or null pointers.
otherwise you will have dangling memory stored there.

for example you want to unset the callbacks and destroy an executor, but here there is still a pointer to the executor

irobot-ros replied:
the whole function could be re-organized a little bit to make it easier to understand.

```
user_callback_data_t * data = &(sub->user_callback_data);

data->callback = callback;
data->user_data = user_data;
data->event_type = SUBSCRIPTION_EVENT;
data->entity_handle = subscription_handle;

if (user_data && subscription_handle && callback) {
    sub->listener = dds_create_listener(data);
    dds_lset_data_on_readers(sub->listener, dds_listener_callback);
    return dds_set_listener(entity_to_listen, sub->listener);
} else {
    return dds_set_listener(entity_to_listen, NULL);
}

```

irobot-ros replied:
Moreover, I'm not sure about the `if (user_data && subscription_handle && callback) {` check.

Should it be with OR instead of AND ?

Then the events executor should check whether pointers are valid or not when it receives stuff.

mauropasse replied:
Regarding comment 1 and 2, OK.
Coment 3, I think is OK to check the 3 pointers are valid here:

1. This is not directly related to events executor, the callback could be something different in the future, so checks have to be done here.
2. The `user_data`, `subscription_handle` and `callback` are used in `dds_listener_callback`, so if I do:
`if (user_data || subscription_handle ||  callback)` , maybe `user_data` and `subscription_handle` are valid but `callback == nullptr` which will cause a seg fault when `dds_listener_callback` is called
3. Checking valid stuff in the events executor every time something is pushed could bring performance degradations, so better check them all here, only once

irobot-ros replied:
sorry, my suggestion about the OR was wrong.

What we should do here is to only check if `callback` is not a null pointer.
That's the only item that is used in the RMW layer, while the others are simply forwarded.

These callbacks are not the events executor callback, they are generic functions that are planned to be used for different use-cases.
Some use-cases will definitely have a null `user_data`.
In theory a callback could also not have any `subscription_handle` (I think they always should, but it's a different story).

The events executor should always check that the pointer it receives are valid.
Checking a couple of pointers to be valid will not cause any performance degradation.
They come from a completely different, RMW-specific, repository, so there is no control on what is put there.

mauropasse replied:
Okoko sounds better checking just the callback.
Regarding `user_data` should be check on `push_event` API of the `EventsExecutor`
```
  static void
  push_event(const void * executor_ptr, rmw_listener_event_t event)
  {
    if(!executor_ptr) return; // Or throw a runtime error
 ```
and `subscription_handle` is already checked on
```
EventsExecutorEntitiesCollector::get_subscription(const void * subscription_id)
{
  auto it = weak_subscriptions_map_.find(subscription_id); 
  // If not found, returns nullptr
 ```

irobot-ros commented on rmw_cyclonedds_cpp/src/rmw_node.cpp:
```diff
@@ -449,22 +467,64 @@ extern "C" rmw_ret_t rmw_set_log_severity(rmw_log_severity_t severity)
+  // 1. The DDS subscription's listener callback.
+  // 2. The user callback, called from the listener callback.
+  /*
+      dds_listener_callback(..) {
```
I find this comment confusing.
this API only set a single thing which is an entity callback provided by the user.
what you describe is how cyclone does that, don't think is relevant here

mauropasse replied:
removed in 010fb92

irobot-ros commented on rmw_cyclonedds_cpp/src/rmw_node.cpp:
```diff
@@ -510,39 +620,111 @@ extern "C" rmw_ret_t rmw_guard_condition_set_listener_callback(
+  // First, lets check if we don't have already a listener.
+
+  // Note: using dds_get_listener(..) when there's no listener
+  // results in a segfault.
```
this is strange, see the code of the function:
```
dds_return_t dds_get_listener (dds_entity_t entity, dds_listener_t *listener)
{
  dds_entity *e;
  dds_return_t ret;
  if (listener == NULL)
    return DDS_RETCODE_BAD_PARAMETER;
  else if ((ret = dds_entity_pin (entity, &e)) != DDS_RETCODE_OK)
    return ret;
  else
  {
    ddsrt_mutex_lock (&e->m_observers_lock);
    dds_copy_listener (listener, &e->m_listener);
    ddsrt_mutex_unlock (&e->m_observers_lock);
    dds_entity_unpin (e);
    return DDS_RETCODE_OK;
  }
}
```

mauropasse replied:
The issue was on `dds_copy_listener (listener, &e->m_listener);` 
Anyway this was just a way to not assign listeners to events, because they override the subscriptions listeners.
The idea is ask how to have different entities for events and subscriptions, then the check for listeners already assigned should dissapear

mauropasse replied:
Just as you say in other comment, `m_listener` is the listener of the entity. Which when you didn't assign a listener before, is a pointer pointing to nowhere, which could cause the issue in the operation `dds_copy_listener` when dereferencing the not valid ptr:
```
void dds_copy_listener (dds_listener_t * __restrict dst, const dds_listener_t * __restrict src)
{
  if (dst && src)
    *dst = *src;
}
```

irobot-ros commented on rmw_cyclonedds_cpp/src/rmw_node.cpp:
```diff
@@ -3193,6 +3375,18 @@ extern "C" rmw_ret_t rmw_trigger_guard_condition(
   dds_set_guardcondition(gcond_impl->gcondh, true);
+
+  // Hack: Get and call the guard condition's listener callback,
+  // as it doesn't seem to be listening to the proper enitity.
```
From a quick look I think that the problem is that the `dds_entity_t` listener field (named `m_listener`) is only used in the `dds_reader_data_available_cb` function.

I think that it is not an hack to do that here, but there are some changes required.
 - you should check the return type of `dds_set_guardcondition`
 - you should do the whole procedure that is currently done in `dds_reader_data_available_cb`
 
 something like this
 ```
 struct dds_listener const * const lst = &rd->m_entity.m_listener;
  dds_entity * const sub = rd->m_entity.m_parent;
  if (lst->on_data_on_readers)
  {
    ddsrt_mutex_unlock (&rd->m_entity.m_observers_lock);
    ddsrt_mutex_lock (&sub->m_observers_lock);
    const uint32_t data_on_rds_enabled = (ddsrt_atomic_ld32 (&sub->m_status.m_status_and_mask) & (DDS_DATA_ON_READERS_STATUS << SAM_ENABLED_SHIFT));
    if (data_on_rds_enabled)
    {
      sub->m_cb_pending_count++;
      while (sub->m_cb_count > 0)
        ddsrt_cond_wait (&sub->m_observers_cond, &sub->m_observers_lock);
      sub->m_cb_count++;
      ddsrt_mutex_unlock (&sub->m_observers_lock);

      lst->on_data_on_readers (sub->m_hdllink.hdl, lst->on_data_on_readers_arg);

      ddsrt_mutex_lock (&sub->m_observers_lock);
      sub->m_cb_count--;
      sub->m_cb_pending_count--;
      ddsrt_cond_broadcast (&sub->m_observers_cond);
    }
    ddsrt_mutex_unlock (&sub->m_observers_lock);
    ddsrt_mutex_lock (&rd->m_entity.m_observers_lock);
  }
  else if (lst->on_data_available)
  {
    ddsrt_mutex_unlock (&rd->m_entity.m_observers_lock);
    lst->on_data_available (rd->m_entity.m_hdllink.hdl, lst->on_data_available_arg);
    ddsrt_mutex_lock (&rd->m_entity.m_observers_lock);
  }
  ```

mauropasse replied:
I have to dig a little more on this, but I think that if you're allowed to attach a listener to an entity, whatever happens with that entity that changes the state of it, should call the listener callback

mauropasse replied:
And yes, seems that `m_listener` is only used in the `dds_reader_data_available_cb` function.
I'll try to implement it here and then we can ask the guys if that's the correct approach

mauropasse replied:
Well seems not straight forward to implement it here. There are no readers, the guard condition is just a single entity.
What I can do is lock the mutex of the guard condition as done in `dds_set_guardcondition` , get the callback, copy it to a new one. Unlock, and call the new one. 
Or just leave it as it is now and ask for guidance?

Commit:
Use dds_entity_t instead of auto. Remove comments

Commit:
PR suggestions

Commit:
Only check callback pointer validity

This is the only item that is used in the RMW
layer, while the others are simply forwarded.

Commit:
use void * to pass executor ptr

Commit:
Implement use previous events for guard conditions

Commit:
Push events happened before setting callback

Commit:
Add support for events unread count

Commit:
Rework unread count

Commit:
Rework all

Commit:
Proper init gc, clients and services callbacks

Commit:
Remove not needed subscription listener init
